### PR TITLE
PN2V Lightning Draft

### PR DESCRIPTION
**Pull request by CatEek**
## Description

> [!NOTE]  
> **tldr**: 

Necessary updates for PN2V

### Overview - what changed?

Added config, loss function, UNet architecture updates and training/predict steps for pn2v

## Changes Made

- `create_pn2v_configuration` added to `configuration_factories.py`
- `pn2v_loss` added to `fcn/losses.py`
- `convert_outputs_pn2v` added to `prediction_utils/prediction_outputs.py`
- `PN2VAlgorithm` added to `config/algorithms.py`

FYI, some of the changes are relevant to fixing pre-commits
## How has this been tested?

Added some tests, performance tested both for pn2v and n2v with careamics-examples 

---

**Please ensure your PR meets the following requirements:**

- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [ ] Pre-commit passes
- [ ] PR to the documentation exists (for bug fixes / features)

### CODE REVIEW

File: src/careamics/config/configuration.py [hunk 1]
```diff
@@ -203,8 +205,8 @@ def validate_n2v_mask_pixel_perc(self: Self) -> Self:
             If the probability of masking a pixel within a patch is less than 1 for the
             chosen masked pixel percentage and patch size.
         """
-        # No validation needed for non n2v algorithms
-        if not isinstance(self.algorithm_config, N2VAlgorithm):
+        # No validation needed for non n2v algorithms # TODO: why ?
```
Comment (jdeschamps) [inline src/careamics/config/configuration.py:208]: To answer the TODO: non N2V algorithms (and non PN2V) do not have n2v mask pixel percentage, so you can't validate it.

File: src/careamics/config/configuration_factories.py [hunk 2]
```diff
@@ -2097,3 +2097,369 @@ def create_microsplit_configuration(
...
+    patch_size: Sequence[int],
+    batch_size: int,
+    nm_path: str,
+    num_epochs: int = 100,
+    num_steps: int | None = None,
+    augmentations: list[Union[XYFlipModel, XYRandomRotate90Model]] | None = None,
+    independent_channels: bool = True,
+    use_n2v2: bool = False,
+    num_in_channels: int = 1,
+    num_out_channels: int = 100,
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py:2133]: What happens when there are indeed multiple channels in the data?

File: src/careamics/config/configuration_factories.py [hunk 3]
```diff
@@ -2097,3 +2097,369 @@ def create_microsplit_configuration(
...
+        ),
+        roi_size=roi_size,
+        masked_pixel_percentage=masked_pixel_percentage,
+        struct_mask_axis=struct_n2v_axis,
+        struct_mask_span=struct_n2v_span,
+    )
+
+    # Create noise model configuration
+    noise_model_config = GaussianMixtureNMConfig(
+        model_type="GaussianMixtureNoiseModel",
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py]: Is there a reason for the literal-typed `model_type` to not have a default value? As it only allows one value, that should be its default one. Then no need to specify it!

File: src/careamics/config/configuration_factories.py [hunk 4]
```diff
@@ -2097,3 +2097,369 @@ def create_microsplit_configuration(
...
+    if num_epochs is not None:
+        final_trainer_params["max_epochs"] = num_epochs
+    if num_steps is not None:
+        final_trainer_params["limit_train_batches"] = num_steps
+
+    training_params = _create_training_configuration(
+        trainer_params=final_trainer_params,
+        logger=logger,
+        checkpoint_params=checkpoint_params,
+    )
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py:2469]: Since these are used now in multiple function, it should be refactored in a new function.

File: src/careamics/config/support/supported_algorithms.py [hunk 5]
```diff
@@ -34,3 +34,6 @@ class SupportedAlgorithm(str, BaseEnum):
 
     HDN = "hdn"
     """Hierarchical Denoising Network, an unsupervised denoising algorithm"""
+
+    PN2V = "pn2v"
+    """Pixel-wise Noise2Void, a self-supervised approach based on blind denoising."""
```
Comment (jdeschamps) [inline src/careamics/config/support/supported_algorithms.py]: Good try Claude!

```suggestion
    """Probabilistic Noise2Void, a probabilistic self-supervised approach based on Noise2Void."""
```

File: src/careamics/lightning/callbacks/data_stats_callback.py [hunk 6]
```diff
@@ -7,21 +7,21 @@
...
         Parameters
         ----------
-        trainer : Lightning.Trainer
-            The trainer instance.
-        module : Lightning.LightningModule
-            The model being trained.
+        trainer : L.Trainer
+            PyTorch Lightning trainer.
+        module : L.LightningModule
+            Lightning module.
```
Comment (jdeschamps) [inline src/careamics/lightning/callbacks/data_stats_callback.py]: The decision a while back was to write out the library it comes from. Although `Lightning` is not the canonical library, it is much more informative than `L`.

```suggestion
        trainer : Lightning.Trainer
            PyTorch Lightning trainer.
        module : Lightning.LightningModule
            Lightning module.
```

File: src/careamics/lightning/lightning_module.py [hunk 7]
```diff
@@ -147,6 +218,12 @@ def training_step(self, batch: torch.Tensor, batch_idx: Any) -> Any:
             aux = []
 
         out = self.model(x_preprocessed)
+
+        # PN2V needs denormalized output and targets for loss computation
+        if isinstance(self.algorithm_config, PN2VAlgorithm):
+            out = self._train_denormalize(out)
```
Comment (jdeschamps) [inline src/careamics/lightning/lightning_module.py:224]: Can you look into the math of the loss function and tell me it would yield a different result if we would normalize the noise model, rather than denormalize the input, and why?

File: src/careamics/nm_training_placeholder.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    self.untrained_noise_models = None
+    if (
+        hasattr(self.cfg.algorithm_config, "train_noise_model")
+        and self.cfg.algorithm_config.train_noise_model_from_data
+    ):
+        self._initialize_noise_models_for_training()
+
+
+# In src/careamics/careamist.py
+def train_noise_model(
```
Comment (jdeschamps) [inline src/careamics/nm_training_placeholder.py:58]: Note for integration later: I would not put this method in CAREamist, but in another module. It is long enough as it is.

A module for NM training would be useful, also for the Lightning API folks.

File: src/careamics/prediction_utils/prediction_outputs.py [hunk 9]
```diff
@@ -41,15 +41,71 @@ def convert_outputs(predictions: list[Any], tiled: bool) -> list[NDArray]:
...
+
+        # Process mmse
+        mse_comb = combine_batches(mse_with_tiles, tiled)
+        mse_output = stitch_prediction(*mse_comb)
+
+        return predictions_output, mse_output
+    else:
+        # Separate predictions and mmse for non-tiled case
+        pred_only = [pred for (pred, mse) in predictions]
+        mse_only = [mse for (pred, mse) in predictions]
```
Comment (jdeschamps) [inline src/careamics/prediction_utils/prediction_outputs.py]: Would that work?

```suggestion
        pred_only, mse_only = zip(*predictions)
```

File: src/careamics/prediction_utils/prediction_outputs.py [hunk 10]
```diff
@@ -41,15 +41,71 @@ def convert_outputs(predictions: list[Any], tiled: bool) -> list[NDArray]:
...
+        return [], []
+
+    if tiled:
+        # Separate predictions and mmse, keeping tile info for each
+        pred_with_tiles = [
+            (pred, tile_info_list) for (pred, _), tile_info_list in predictions
+        ]
+        mse_with_tiles = [
+            (mse, tile_info_list) for (_, mse), tile_info_list in predictions
+        ]
```
Comment (jdeschamps) [inline src/careamics/prediction_utils/prediction_outputs.py:78]: I would prefer the more straightforward for loop, for readability pruposes

```suggestion
pred_with_tiles=[]
mse_with_tiles = []
for (pred, mse), tile_info_list in predictions:
    pred_with_tiles.append((pred, tile_info_list))
    mse_with_tiles.append((mse, tile_info_list))
```


Alternatively, if you really love list iterations that make people read the code four times, this may work:

```
pred_with_tiles, mse_with_tiles = zip(
    *[((pred, tile_info_list), (mse, tile_info_list)) for (pred, mse), tile_info_list in predictions]
)
pred_with_tiles = list(pred_with_tiles)
mse_with_tiles = list(mse_with_tiles)
```

File: tests/losses/test_pn2v_loss.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    samples = torch.randn(batch_size, 1, height, width)
+    labels = torch.randn(batch_size, 1, height, width)
+    masks = torch.randint(0, 2, (batch_size, 1, height, width)).float()
+
+    # Create a noise model config
+    nm_config = GaussianMixtureNMConfig(
+        model_type="GaussianMixtureNoiseModel", path=tmp_path / "dummy_noise_model.npz"
+    )
+
+    # Create noise model (we'll need to mock this properly)
```
Comment (jdeschamps) [inline tests/losses/test_pn2v_loss.py]: Can you add a TODO?

**Review by jdeschamps — CHANGES_REQUESTED** [review]
Thanks!

Some minor changes and requests. But my main concern is what happens when there are multiple channels in the raw image in PN2V?

I am also not convinced yet by the noise model pipeline. For now it would be ok, but we definitely need to sit down and discuss it. I'd love to understand it more.

File: src/careamics/config/configuration_factories.py [hunk 12]
```diff
@@ -2097,3 +2097,369 @@ def create_microsplit_configuration(
...
+        raise ValueError("num_in_channels must be at least 1 when using channels.")
+    elif "C" not in axes and num_in_channels > 1:
+        raise ValueError(
+            f"C is not present in the axes, but num_in_channels is specified "
+            f"(got {num_in_channels} channels)."
+        )
+
+    # Calculate total output channels based on independent_channels setting
+    if independent_channels:
+        total_out_channels = num_in_channels * num_out_channels
```
Comment (CatEek) [inline src/careamics/config/configuration_factories.py:2403]: They will be splitted in the output, but we haven't tested and we don't have a use case so far

File: src/careamics/config/configuration_factories.py (hunk 4 above)
Comment (CatEek) [reply src/careamics/config/configuration_factories.py:2469]: Not sure I follow

File: src/careamics/config/configuration_factories.py (hunk 2 above)
Comment (jdeschamps) [reply src/careamics/config/configuration_factories.py:2133]: Given the fact that the number of model output channels is scaled by the number of actual biological channels, these need to be reconciled during prediction into two averaged channels.

Look at the prediction, leave a TODO or fix. But if you do a TODO, it needs to be taken care of in the next PR.

File: src/careamics/config/configuration_factories.py (hunk 4 above)
Comment (jdeschamps) [reply src/careamics/config/configuration_factories.py:2469]: Leave a TODO to refactor this bit of code away from the various convenience functions, because it is duplicated throughout.

If you don't do this in this PR, because it is out of scope, please to a quick PR about it soon-ish!

Comment (CatEek) [reply src/careamics/config/configuration_factories.py:2469]: Done

File: src/careamics/lightning/lightning_module.py (hunk 7 above)
Comment (CatEek) [reply src/careamics/lightning/lightning_module.py:224]: Will do later

File: src/careamics/prediction_utils/prediction_outputs.py (hunk 10 above)
Comment (jdeschamps) [reply src/careamics/prediction_utils/prediction_outputs.py:78]: Actually I can see the problem with the indentation here, probably the problem was on my end (my code suggestion did not have the indentation).

But you reverted to the previous code snippet?

File: tests/losses/test_pn2v_loss.py (hunk 11 above)
Comment (jdeschamps) [reply tests/losses/test_pn2v_loss.py]: You thumbed up, but I don't see the todo

File: src/careamics/config/configuration_factories.py (hunk 12 above)
Comment (jdeschamps) [reply src/careamics/config/configuration_factories.py:2403]: Can you add a TODO?

File: tests/losses/test_pn2v_loss.py (hunk 11 above)
Comment (CatEek) [reply tests/losses/test_pn2v_loss.py]: Done

**Review by jdeschamps — APPROVED** [review]

**Comment by jdeschamps** [comment]
You have pre-commit errors.

Also bear in mind that we will need to implement quite soon a PN2V Lightning Module compatible with the NG Dataset. While it is still fresh in your memory, it might make sense to do it soon.
